diff --git a/core/modules/system/src/Controller/AssetControllerBase.php b/core/modules/system/src/Controller/AssetControllerBase.php
index 4f8d5ad6965..101bea2359b 100644
--- a/core/modules/system/src/Controller/AssetControllerBase.php
+++ b/core/modules/system/src/Controller/AssetControllerBase.php
@@ -2,6 +2,7 @@
 
 namespace Drupal\system\Controller;
 
+use Drupal\Component\Render\FormattableMarkup;
 use Drupal\Component\Utility\UrlHelper;
 use Drupal\Core\Asset\AssetCollectionGrouperInterface;
 use Drupal\Core\Asset\AssetCollectionOptimizerInterface;
@@ -11,6 +12,7 @@
 use Drupal\Core\Asset\AttachedAssets;
 use Drupal\Core\Asset\AttachedAssetsInterface;
 use Drupal\Core\Asset\LibraryDependencyResolverInterface;
+use Drupal\Core\File\FileUrlGeneratorInterface;
 use Drupal\Core\StreamWrapper\StreamWrapperManagerInterface;
 use Drupal\Core\Theme\ThemeInitializationInterface;
 use Drupal\Core\Theme\ThemeManagerInterface;
@@ -92,6 +94,7 @@ public function __construct(
     protected readonly AssetCollectionGrouperInterface $grouper,
     protected readonly AssetCollectionOptimizerInterface $optimizer,
     protected readonly AssetDumperUriInterface $dumper,
+    protected readonly FileUrlGeneratorInterface $fileUrlGenerator,
   ) {
     parent::__construct($streamWrapperManager);
     $this->fileExtension = $this->assetType;
@@ -205,9 +208,18 @@ public function deliver(Request $request, string $file_name) {
       $this->dumper->dumpToUri($data, $this->assetType, $uri);
     }
     else {
+      file_put_contents(DRUPAL_ROOT . '/public/sites/default/files/debug.log', (string) new FormattableMarkup('Hash validation failed for [@asset]. Expected hash [@expected_hash], got [@actual_hash]. Group: @group', [
+        '@asset' => $uri,
+        '@expected_hash' => $generated_hash,
+        '@actual_hash' => $received_hash,
+        '@group' => print_r($group, TRUE),
+      ]), FILE_APPEND);
       $expected_filename = $this->fileExtension . '_' . $generated_hash . '.' . $this->fileExtension;
+      $uri = 'assets://' . $this->fileExtension . '/' . $expected_filename;
+      $parsed_url = UrlHelper::parse($request->getRequestUri());
+
       $response = new RedirectResponse(
-        str_replace($file_name, $expected_filename, $request->getRequestUri()),
+        $this->fileUrlGenerator->generateString($uri) . '?' . UrlHelper::buildQuery($parsed_url['query'] ?? []),
         301,
         ['Cache-Control' => 'public, max-age=3600, must-revalidate'],
       );
diff --git a/core/modules/system/src/Controller/CssAssetController.php b/core/modules/system/src/Controller/CssAssetController.php
index 76d1e8a8b71..5051510bcda 100644
--- a/core/modules/system/src/Controller/CssAssetController.php
+++ b/core/modules/system/src/Controller/CssAssetController.php
@@ -38,6 +38,7 @@ public static function create(ContainerInterface $container) {
       $container->get('asset.css.collection_grouper'),
       $container->get('asset.css.collection_optimizer'),
       $container->get('asset.css.dumper'),
+      $container->get('file_url_generator'),
     );
   }
 
diff --git a/core/modules/system/src/Controller/JsAssetController.php b/core/modules/system/src/Controller/JsAssetController.php
index 221f53410dd..1154d9368a8 100644
--- a/core/modules/system/src/Controller/JsAssetController.php
+++ b/core/modules/system/src/Controller/JsAssetController.php
@@ -38,6 +38,7 @@ public static function create(ContainerInterface $container) {
       $container->get('asset.js.collection_grouper'),
       $container->get('asset.js.collection_optimizer'),
       $container->get('asset.js.dumper'),
+      $container->get('file_url_generator'),
     );
   }
 
