diff --git a/draggableviews.module b/draggableviews.module
--- a/draggableviews.module
+++ b/draggableviews.module
@@ -179,6 +179,7 @@
 
   $connection = Database::getConnection();
   $transaction = $connection->startTransaction();
+  $langcode = \Drupal::languageManager()->getCurrentLanguage(\Drupal\Core\Language\LanguageInterface::TYPE_CONTENT)->getId();
   try {
     foreach ($input['draggableviews'] as $item) {
       // Remove old data.
@@ -187,6 +188,7 @@
         ->condition('view_display', $view_display)
         ->condition('args', $view_args)
         ->condition('entity_id', $item['id'])
+        ->condition('langcode', $langcode)
         ->execute();
 
       // Add new data.
@@ -196,6 +198,7 @@
         'args' => $view_args,
         'entity_id' => $item['id'],
         'weight' => $item['weight'],
+        'langcode' => $langcode,
       ];
       // Save parent if exists.
       if (isset($item['parent'])) {
